### PR TITLE
Improve performance of cudf::strings::split on whitespace

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Improves performance of strings split on whitespace by reducing character counting where possible/practical.
The gbench benchmark was recoded to use nvbench.
Some code was refactored for better reuse with split and split-record.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
Performance results showed the highest improvement for longer strings
```
| width |   rows   |   Ref Time |  New Time  |          Diff |   %Diff |
|-------|----------|------------|------------|---------------|---------|
|  32   |   4096   | 230.359 us | 214.736 us |    -15.622 us |  -6.78% |
|  64   |   4096   | 304.666 us | 282.046 us |    -22.620 us |  -7.42% |
|  128  |   4096   | 445.859 us | 413.572 us |    -32.287 us |  -7.24% |
|  256  |   4096   | 696.039 us | 633.025 us |    -63.014 us |  -9.05% |
|  512  |   4096   |   1.170 ms |   1.046 ms |   -124.532 us | -10.64% |
| 1024  |   4096   |   2.235 ms |   1.863 ms |   -372.772 us | -16.68% |
| 2048  |   4096   |   4.444 ms |   3.652 ms |   -792.256 us | -17.83% |
|  32   |  32768   | 274.620 us | 261.166 us |    -13.454 us |  -4.90% |
|  64   |  32768   | 333.561 us | 312.561 us |    -20.999 us |  -6.30% |
|  128  |  32768   | 516.326 us | 481.860 us |    -34.465 us |  -6.68% |
|  256  |  32768   | 834.735 us | 784.460 us |    -50.275 us |  -6.02% |
|  512  |  32768   |   1.395 ms |   1.299 ms |    -95.756 us |  -6.87% |
| 1024  |  32768   |   2.591 ms |   2.289 ms |   -302.364 us | -11.67% |
| 2048  |  32768   |   5.544 ms |   4.307 ms |  -1236.045 us | -22.30% |
|  32   |  262144  | 508.737 us | 489.034 us |    -19.702 us |  -3.87% |
|  64   |  262144  | 746.182 us | 705.883 us |    -40.298 us |  -5.40% |
|  128  |  262144  |   1.478 ms |   1.382 ms |    -95.135 us |  -6.44% |
|  256  |  262144  |   3.672 ms |   3.282 ms |   -390.133 us | -10.62% |
|  512  |  262144  |   7.404 ms |   6.332 ms |  -1072.387 us | -14.48% |
| 1024  |  262144  |  16.137 ms |  12.686 ms |  -3451.053 us | -21.39% |
| 2048  |  262144  |  40.868 ms |  28.841 ms | -12026.713 us | -29.43% |
|  32   | 2097152  |   2.331 ms |   2.249 ms |    -81.614 us |  -3.50% |
|  64   | 2097152  |   3.839 ms |   3.629 ms |   -209.851 us |  -5.47% |
|  128  | 2097152  |   9.143 ms |   8.641 ms |   -501.609 us |  -5.49% |
|  256  | 2097152  |  31.479 ms |  26.316 ms |  -5162.946 us | -16.40% |
|  512  | 2097152  |  69.241 ms |  55.764 ms | -13476.870 us | -19.46% |
|  32   | 16777216 |  16.374 ms |  15.840 ms |   -533.659 us |  -3.26% |
|  64   | 16777216 |  27.919 ms |  26.359 ms |  -1559.420 us |  -5.59% |
```

File: cpp/src/strings/split/split_record.cu [hunk 1]
```diff
@@ -79,50 +79,24 @@ std::unique_ptr<column> split_record_fn(strings_column_view const& input,
...
-  }
-};
-
 /**
  * @brief Identify the tokens from the `idx'th` string element of `d_strings`.
  */
 template <Dir dir>
 struct whitespace_token_reader_fn {
   column_device_view const d_strings;  // strings to split
   size_type const max_tokens{};
```
Comment (karthikeyann) [inline cpp/src/strings/split/split_record.cu:88]: default value of this `max_tokens` and `whitespace_token_counter_fn::max_tokens` are different.
could we make both of them same?

File: cpp/src/strings/split/split_record.cu [hunk 2]
```diff
@@ -79,50 +79,24 @@ std::unique_ptr<column> split_record_fn(strings_column_view const& input,
...
-    // pandas.Series.str.split("") returns 0 tokens.
-    if (reached_max_tokens || !spaces) token_count++;
-    return token_count;
-  }
-};
-
 /**
  * @brief Identify the tokens from the `idx'th` string element of `d_strings`.
  */
 template <Dir dir>
```
Comment (karthikeyann) [inline cpp/src/strings/split/split_record.cu:85]: prefer to use full name for template variable. `direction`

**Review by karthikeyann — APPROVED** [review]

File: cpp/benchmarks/string/split.cpp [hunk 3]
```diff
@@ -25,63 +23,49 @@
...
+               [&](nvbench::launch& launch) { cudf::strings::split(input, target); });
+  } else if (stype == "split_ws") {
+    state.exec(nvbench::exec_tag::sync,
+               [&](nvbench::launch& launch) { cudf::strings::split(input); });
+  } else if (stype == "record") {
+    state.exec(nvbench::exec_tag::sync,
+               [&](nvbench::launch& launch) { cudf::strings::split_record(input, target); });
+  } else {
+    state.exec(nvbench::exec_tag::sync,
+               [&](nvbench::launch& launch) { cudf::strings::split_record(input); });
```
Comment (ttnghia) [inline cpp/benchmarks/string/split.cpp:63]: Nit: We can use just one `state.exec()` call with the `if/else` statements inside its lambda.

File: cpp/include/cudf/strings/detail/split_utils.cuh [hunk 4]
```diff
@@ -22,6 +22,31 @@ namespace cudf {
 namespace strings {
 namespace detail {
 
+constexpr bool is_whitespace(char_utf8 ch) { return ch <= ' '; }
```
Comment (ttnghia) [inline cpp/include/cudf/strings/detail/split_utils.cuh:25]: Why does this checking "is whitespace" return true for all characters `<= ' '`? Why not just `ch == ' '`?

File: cpp/include/cudf/strings/detail/split_utils.cuh [hunk 5]
```diff
@@ -22,6 +22,31 @@ namespace cudf {
...
+        itr += chr_width;
+      } else {
+        token_count += static_cast<size_type>(spaces);
+        spaces = !spaces;
+      }
+    }
+    return token_count;
+  }
+
+  size_type const max_tokens = std::numeric_limits<size_type>::max();
```
Comment (ttnghia) [inline cpp/include/cudf/strings/detail/split_utils.cuh:47]: Should this be `constexpr`?

File: cpp/include/cudf/strings/detail/split_utils.cuh [hunk 6]
```diff
@@ -22,6 +22,31 @@ namespace cudf {
 namespace strings {
 namespace detail {
 
+constexpr bool is_whitespace(char_utf8 ch) { return ch <= ' '; }
+
+struct whitespace_token_counter_fn {
+  __device__ size_type count_tokens(string_view d_str) const
```
Comment (ttnghia) [inline cpp/include/cudf/strings/detail/split_utils.cuh:28]: I see a little bit awkward when using this struct:
```
  whitespace_token_counter_fn counter{max_tokens};
  return counter.count_tokens(d_str);
```

Why not just make this a device function like `count_whitespace_token(...)`? Then we can just use this:
```
  return count_whitespace_token(d_str, max_tokens);
```

Comment (davidwendt) [reply cpp/include/cudf/strings/detail/split_utils.cuh:28]: It needs to maintain state across calls.

File: cpp/include/cudf/strings/detail/split_utils.cuh (hunk 4 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/detail/split_utils.cuh:25]: Space is not the same as whitespace. The `<= ' '` accounts for new-lines, tabs, carriage returns, etc.

File: cpp/benchmarks/string/split.cpp (hunk 3 above)
Comment (davidwendt) [reply cpp/benchmarks/string/split.cpp:63]: I don't want to include the if-statement string compares in the benchmark measurements.

**Pull request merged**